### PR TITLE
Switch to apollo

### PR DESCRIPTION
**Pull request by xavxyz**
~WIP~ 
edit: ⬇️ see below :octocat:

### CODE REVIEW

**Comment by stubailo** [comment]
Oooh, `yarn` - fancy!

File: imports/server/parse.js [hunk 1]
```diff
@@ -12,11 +12,15 @@ export function parseMD(md) {
 
   const contentBlocks = parseContentBlocks(content, metadata);
 
+  // these lines have already been used to create the introduction step,
+  // remove them from the step root
+  const { lineStart, lineEnd, ...cleanMetadata } = metadata;
```
Comment (xavxyz) [inline imports/server/parse.js]: in reference to #11

File: imports/components/Tour.js [hunk 2]
```diff
@@ -17,65 +14,53 @@ class Tour extends React.Component {
...
+      await this.props.importTour();
+    } catch (error) {
+      alert(error);
+    }
   }
 
   render() {
+    const { location, tour } = this.props;
+
+    const [tourAuthor] = tour.repository.split('/');
```
Comment (xavxyz) [inline imports/components/Tour.js]: I'm starting to simplify the components, more work to do on them later 💯 (💅?)

File: imports/components/Tour.js [hunk 3]
```diff
@@ -149,32 +130,72 @@ class Tour extends React.Component {
...
+        // note: tried without, it's not really updating the ui, maybe it's because
+        // of how we are doing our mutation?
+        refetchQueries: ['getTourWithSteps'],
+      }),
+    }),
+  }
+);
+
+export default compose(withTour, displayLoadingState, displayNotFoundState, withMutation, pure)(
+  Tour
```
Comment (xavxyz) [inline imports/components/Tour.js]: `prettier` 🙄 

It's a really cool tool, but sometimes I feel like it gets crazy (especially with `gql`) -> weird indentation

File: main/client/main.html [hunk 4]
```diff
@@ -15,5 +15,4 @@
 </head>
 
 <body>
-  <div id="root"></div>
```
Comment (xavxyz) [inline main/client/main.html:18]: This one tricked me 🙄 "why the hell do I have two pages?" 😄

File: imports/server/index.js [hunk 5]
```diff
@@ -12,11 +13,21 @@ const schema = makeExecutableSchema({
...
+    context: {
+      github: new GitHubConnector(),
+      Tours,
+      Steps,
+      opticsContext: OpticsAgent.context(req),
+    },
+  }),
+  {
+    configServer: server => {
+      server.use('/graphql', OpticsAgent.middleware());
```
Comment (xavxyz) [inline imports/server/index.js:30]: 🙃 https://cl.ly/3g023M2k3W3x

File: imports/components/Tour.js [hunk 6]
```diff
@@ -149,32 +130,72 @@ class Tour extends React.Component {
...
+  }
+`,
+  {
+    props: ({ ownProps: { tour: { repository } }, mutate }) => ({
+      importTour: () => mutate({
+        variables: { tourRepository: repository },
+        // basic *forced* update, we shouldn't do that normally (automatic update)
+        // note: tried without, it's not really updating the ui, maybe it's because
+        // of how we are doing our mutation?
+        refetchQueries: ['getTourWithSteps'],
```
Comment (xavxyz) [inline imports/components/Tour.js]: 💡 makes sense! we remove and then insert the tour, so the `_id` changes and so the store freaks out! 🌮 🎉  fixing that.

**Comment by xavxyz** [comment]
🎈 Ready to take off
---
We have a fancy GraphQL layer handling our tours & steps 💃 

Server-side capabilities
---
- [x] Queries
- [x] Mutations
- [x] Types
- [x] Server-side rendering with `renderDataToString` (including headtags \o/)
- [x] Wired to Optics

Data loading from the UI
---
- [x] Containers loading data over DDP replaced by containers loading data with `graphql` HOC: search recent tours, load a tour and its steps, load a step
- [x] Import tour mutation
- [x] Moving between steps (`prev` & `next` in the schema)

Next steps (by documenting ideas/issues)
---
We have two types of users: the writer ✍️  and the reader 👀 (..... who can become a 🗣!) We have room for a lot of crazy interesting stuff here! 🚀 

On the ✍️ experience, we could improve how we parse files and provide more guidance / helpful errors / more automatic flow (#1) ? Also supporting images could be great (#2)
Edit: prefetching steps when hovering the button, transitions between routes / components re-loading, good way to display errors 

On the 👀 experience, we could improve the navigation between steps with key controls, making sections/code easier to read (#6).

➡️ All this to lead to an awesome code documentation / presentation tool!

*And in fact there is another kind of users: the contributors  ~:party_parrot:~* :octocat:

When this PR is merged I'm really into adding storybook and simplify components. In this process, I'm in favor of using `styled-components`:
- [theming](https://github.com/styled-components/styled-components#theming), you could theme your tour!
- colocate your styles with your components, among the benefits: readability, real CSS with the power of JS, no more importing tons of LESS files for bootstrap (lol), no more inline-styles

I'm thinking also about:
- batching requests server-side with `dataloader` for the technical experiments ; 
- add tests on the client with storyshots (jest snapshots fusions w/ storybook)
- add tests on mongo / github connector.
- use fragments for our queries (and maybe `.graphql` files?)

Yay, here are the things that comes through my head at the moment 🦆 (emoji the closest to a parrot)!

What do you think?

**Comment by xavxyz** [comment]
Shall I merge dat? I've already started something for the new components on [simpler-components branch](https://github.com/partyparrot/codetours/tree/simpler-components)

**Comment by stubailo** [comment]
@xavcz hey, this looks awesome - going to merge and deploy. Sorry I've been slow on this, got hit like a truck by the Apollo 1.0 release :]

**Comment by stubailo** [comment]
@xavcz BTW I set up Optics on the `partyparrot` org so you should be able to see it here, looks like it's pretty fast! 🎉  https://optics.apollodata.com/service/partyparrot-codetours

**Comment by stubailo** [comment]
I did have to re-import all of the tours after deploying, but it seems to work fine now :]

**Comment by xavxyz** [comment]
🎉🎉 that rocks! 

No bother for the wait, congrats to you and the team for delivering it!! Great job 🎩

I have commit right but I'm not in the party parrot organization, can you add me please? 😁

**Comment by stubailo** [comment]
Done!

**Pull request merged**